src/main/java/reposense/report/ReportGenerator.java

```diff
@@ -46,34 +50,75 @@ public static void generateReposReport(List<RepoConfiguration> configs, String o
+     */
+    private static Map<RepoLocation, List<RepoConfiguration>> groupConfigsByRepoLocation(
+            List<RepoConfiguration> configs) {
+        Map<RepoLocation, List<RepoConfiguration>> repoLocationMap = new HashMap<>();
+        for (RepoConfiguration config : configs) {
+            RepoLocation location = config.getLocation();
+            if (repoLocationMap.containsKey(location)) {
+                repoLocationMap.get(location).add(config);
+            } else {
+                List<RepoConfiguration> configsList = new ArrayList<>();
+                configsList.add(config);
+                repoLocationMap.put(location, configsList);
+            }
+        }
+        return repoLocationMap;
+    }
+
+    /**
+     * Clone, analyze and generate the report for repositories in {@code repoLocationMap}.
      * Performs analysis and report generation of each repository in parallel with the cloning of the next repository.
      */
-    private static void cloneAndAnalyzeRepos(List<RepoConfiguration> configs, String outputPath) throws IOException {
+    private static void cloneAndAnalyzeRepos(
+            Map<RepoLocation, List<RepoConfiguration>> repoLocationMap, String outputPath) throws IOException {
         RepoCloner repoCloner = new RepoCloner();
-        RepoConfiguration clonedRepo = null;
+        RepoLocation clonedRepoLocation = null;
 
-        for (RepoConfiguration config : configs) {
-            repoCloner.clone(outputPath, config);
+        for (RepoLocation location : repoLocationMap.keySet()) {
+            repoCloner.clone(outputPath, repoLocationMap.get(location).get(0));
 
-            if (clonedRepo != null) {
-                analyzeRepo(outputPath, clonedRepo);
+            if (clonedRepoLocation != null) {
+                analyzeRepos(outputPath, repoLocationMap.get(clonedRepoLocation), repoCloner);
             }
-            clonedRepo = repoCloner.getClonedRepo(outputPath);
+            clonedRepoLocation = repoCloner.getClonedRepo(outputPath);
         }
-        if (clonedRepo != null) {
-            analyzeRepo(outputPath, clonedRepo);
+        if (clonedRepoLocation != null) {
+            analyzeRepos(outputPath, repoLocationMap.get(clonedRepoLocation), repoCloner);
         }
         repoCloner.cleanup();
     }
 
+    /**
+     * Analyzes repos in {@code configs} and generates the report.
+     */
+    private static void analyzeRepos(
+            String outputPath, List<RepoConfiguration> configs, RepoCloner repoCloner) throws IOException {
+
+        for (RepoConfiguration config : configs) {
+            try {
+                repoCloner.updateAndCheckoutBranch(config);
+            } catch (RuntimeException e) {
+                logger.log(Level.SEVERE, "Branch does not exist! Analysis terminated.", e);
+                Path repoReportDirectory = Paths.get(outputPath, config.getDisplayName());
+                FileUtil.createDirectory(repoReportDirectory);
```

Comment:
Would it be possible to move the creation of the folder up (think we can create a helper method for this to improve readability for this), and also do the error handling up here as well (if failed to create)? Then, we won't have to create the folder in the other method `analyzeRepo`.

Reply:
Yup that sound better.